# "Hallelujah Effect" Principal Components Analysis

This notebook models the "Hallelujah Effect" in terms of all basic features available in the dataset for those subjects that listened to the song and had an EDA quality >80%.

In [31]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

from sklearn import decomposition

In [32]:
# Set bucket, project, and region
# BUCKET = 'eim-muse'
# PROJECT = 'eim-muse'
# REGION = 'us-central1'

In [36]:
# import os
# os.environ['BUCKET'] = BUCKET
# os.environ['PROJECT'] = PROJECT
# os.environ['REGION'] = REGION

In [37]:
# %bash
# gcloud config set project $PROJECT
# gcloud config set compute/region $REGION

## Retrieve and Subset Datasource

Get data from BigQuery but defer filtering, etc. to Beam. Data in BigQuery has been pre-processed with Dataprep.

In [ ]:
import google.datalab.bigquery as bq
def create_query():
  base_query = """
SELECT *
FROM
  `eim-muse.hallelujah_effect.full_hallelujah_trials_cleaned`
  """
    
  return base_query

query = create_query()

In [ ]:
df = bq.Query(query).execute().result().to_dataframe()
df.head()
df.describe()

In [ ]:
df.columns

In [ ]:
df.drop(columns='id')

df['hearing_impairments'] = df['hearing_impairments'].astype(int)

nationality_indicators = pd.get_dummies(df['nationality'], prefix='nationality')
df = df.drop(columns='nationality')
df = pd.concat([df, nationality_indicators])
df.head()

# df['sex'] = pd.get_dummies(df['sex'], prefix='sex')
# df['hallelujah_reaction'] = pd.get_dummies(df['hallelujah_reaction'], prefix='hallelujah_reaction')
# df['location'] = pd.get_dummies(df['location'], prefix='location')
# df['language'] = pd.get_dummies(df['language'], prefix='language')

In [ ]:
df.iloc[0]

In [30]:
df.columns

Index([u'id', u'age', u'concentration', u'hearing_impairments',
       u'musical_expertise', u'nationality', u'artistic', u'fault',
       u'imagination', u'lazy', u'nervous', u'outgoing', u'reserved',
       u'stress', u'thorough', u'trusting', u'activity', u'engagement',
       u'familiarity', u'like_dislike', u'positivity', u'tension', u'sex',
       u'hallelujah_reaction', u'location', u'language', u'music_pref_none',
       u'music_pref_hiphop', u'music_pref_dance', u'music_pref_world',
       u'music_pref_rock', u'music_pref_pop', u'music_pref_classical',
       u'music_pref_jazz', u'music_pref_folk',
       u'music_pref_traditional_irish'],
      dtype='object')

In [8]:
pca = decomposition.PCA()

X = df['age', 'concentration', 'hearing_impairments', 'musical_expertise', u'nationality', u'artistic', u'fault',
       u'imagination', u'lazy', u'nervous', u'outgoing', u'reserved',
       u'stress', u'thorough', u'trusting', u'activity', u'engagement',
       u'familiarity', u'like_dislike', u'positivity', u'tension', u'sex',
       u'hallelujah_reaction', u'location', u'language', u'music_pref_none',
       u'music_pref_hiphop', u'music_pref_dance', u'music_pref_world',
       u'music_pref_rock', u'music_pref_pop', u'music_pref_classical',
       u'music_pref_jazz', u'music_pref_folk',
       u'music_pref_traditional_irish']]

303 training examples / 61 evaluation examples


age  concentration  hearing_impairments  musical_expertise  artistic  \
0      5       2.000000                False            1.00000  3.000000   
1     31       2.000000                False            1.00000  2.000000   
2     21       4.000000                False            3.00000  3.000000   
3     12       5.000000                False            4.00000  2.000000   
4      8       5.000000                False            1.00000  5.000000   
5     32       5.000000                False            3.00000  1.000000   
6     54       5.000000                False            2.00000  2.000000   
7     33       5.000000                False            2.00000  2.000000   
8     17       4.000000                False            1.00000  2.000000   
9      8       5.000000                False            2.00000  5.000000   
10    10       4.000000                False            4.00000  5.000000   
11     9       3.000000                False            2.00000  3.000000   
12    11       2.000000                False            2.00000  1.000000   
13    50       5.000000                False            3.00000  2.000000   
14    11       3.000000                False            3.00000  3.000000   
15    20       4.000000                False            1.00000  2.000000   
16     9       3.000000                False            1.00000  2.000000   
17    14       5.000000                False            3.00000  5.000000   
18    13       3.000000                 True            4.00000  4.000000   
19     7       3.000000                False            2.00000  3.000000   
20    40       4.000000                False            1.00000  3.000000   
21    24       3.000000                False            4.00000  3.000000   
22    13       5.000000                False            3.00000  3.000000   
23    13       5.000000                False            3.00000  3.000000   
24    18       4.000000                False            3.00000  1.000000   
25    10       5.000000                False            1.00000  4.000000   
26    11       5.000000                False            1.00000  5.000000   
27     9       4.000000                False            2.00000  1.000000   
28    12       5.000000                False            2.00000  3.000000   
29    33       2.000000                False            4.00000  1.000000   
..   ...            ...                  ...                ...       ...   
334   21       3.991266                 True            2.52988  2.353712   
335   28       3.991266                False            2.52988  2.353712   
336   43       3.991266                False            2.52988  2.353712   
337   41       3.991266                False            2.52988  2.353712   
338   21       3.991266                False            2.52988  2.353712   
339   21       3.991266                False            2.52988  2.353712   
340   25       3.991266                False            2.52988  2.353712   
341   19       3.991266                False            2.52988  2.353712   
342   28       3.991266                False            2.52988  2.353712   
343   41       3.991266                False            2.52988  2.353712   
344   21       3.991266                False            2.52988  2.353712   
345   34       3.991266                 True            4.00000  2.353712   
346   21       3.991266                False            2.52988  2.353712   
347   32       3.991266                False            2.52988  2.353712   
348   41       3.991266                False            2.52988  2.353712   
349   25       3.991266                False            2.52988  2.353712   
350   21       3.991266                False            2.52988  2.353712   
351   15       3.991266                False            2.52988  2.353712   
352   22       3.991266                False            2.52988  2.353712   
353   51       3.991266                False            2.52988  2.353712   
3

## Create ML dataset using tf.transform and Dataflow

Let's use Cloud Dataflow to read in the BigQuery data and write it out as CSV files. Along the way, let's use tf.transform to do scaling and transforming. Using tf.transform allows us to save the metadata to ensure that the appropriate transformations get carried out during prediction as well.

In [10]:
%writefile requirements.txt
tensorflow-transform==0.6.0

Overwriting requirements.txt


In [15]:
import datetime
import tensorflow as tf
import apache_beam as beam
import tensorflow_transform as tft
from tensorflow_transform.beam import impl as beam_impl

def is_valid(inputs):
    try:
        return True
    except:
        return False

float_features = [
    'activity',
    'age',
    'artistic',
    'concentration',
    'engagement',
    'familiarity',
    'fault',
    'imagination',
    'lazy',
    'like_dislike',
    'musical_expertise',
    'nervous',
    'outgoing',
    'positivity',
    'reserved',
    'stress',
    'tension',
    'thorough',
    'trusting'
]

boolean_features = [
    'hallelujah_reaction',
    'hearing_impairments',
    'music_pref_classical',
    'music_pref_dance',
    'music_pref_folk',
    'music_pref_hiphop',
    'music_pref_jazz',
    'music_pref_none',
    'music_pref_pop',
    'music_pref_rock',
    'music_pref_traditional_irish',
    'music_pref_world'
]

categorical_features = [
    'language',
    'location',
    'nationality',
    'sex'
]

def preprocess_tft(inputs):
    import datetime
    result = {}
    
    for feature in float_features:
        result[feature] = tft.scale_to_0_1(inputs[feature])
    
    for feature in boolean_features:
        result[feature] = tf.cast(inputs[feature], tf.int64)
    
    for feature in categorical_features:
        result[feature] = tf.identity(inputs[feature])
    
    return result

def preprocess(in_test_mode, EVERY_N=None):
  import os
  import os.path
  import tempfile
  from apache_beam.io import tfrecordio
  from tensorflow_transform.coders import example_proto_coder
  from tensorflow_transform.tf_metadata import dataset_metadata
  from tensorflow_transform.tf_metadata import dataset_schema
  from tensorflow_transform.beam import tft_beam_io
  from tensorflow_transform.beam.tft_beam_io import transform_fn_io

  job_name = 'hallelujah-effect-features' + '-' + datetime.datetime.now().strftime('%y%m%d-%H%M%S')    
  if in_test_mode:
    import shutil
    print 'Launching local job ... hang on'
    OUTPUT_DIR = './preproc_tft'
    shutil.rmtree(OUTPUT_DIR, ignore_errors=True)
    
  else:
    print 'Launching Dataflow job {} ... hang on'.format(job_name)
    OUTPUT_DIR = 'gs://{0}/analysis/hallelujah-effect/preproc_tft/'.format(BUCKET)
    import subprocess
    subprocess.call('gsutil rm -r {}'.format(OUTPUT_DIR).split())
  
  # Configure Beam pipeline options
  options = {
    'staging_location': os.path.join(OUTPUT_DIR, 'tmp', 'staging'),
    'temp_location': os.path.join(OUTPUT_DIR, 'tmp'),
    'job_name': job_name,
    'project': PROJECT,
    'max_num_workers': 24,
    'teardown_policy': 'TEARDOWN_ALWAYS',
    'no_save_main_session': True,
    'requirements_file': 'requirements.txt'
  }
  opts = beam.pipeline.PipelineOptions(flags=[], **options)
  if in_test_mode:
    RUNNER = 'DirectRunner'
  else:
    RUNNER = 'DataflowRunner'

  # Setup metadata
  raw_data_schema = {
    colname : dataset_schema.ColumnSchema(tf.string, [], dataset_schema.FixedColumnRepresentation())
                   for colname in categorical_features
  }
  raw_data_schema.update({
      colname : dataset_schema.ColumnSchema(tf.float32, [], dataset_schema.FixedColumnRepresentation())
                   for colname in float_features
    })
  raw_data_schema.update({
      colname : dataset_schema.ColumnSchema(tf.int64, [], dataset_schema.FixedColumnRepresentation())
                   for colname in boolean_features
    })
  raw_data_metadata = dataset_metadata.DatasetMetadata(dataset_schema.Schema(raw_data_schema))
  
  # run Beam  
  with beam.Pipeline(RUNNER, options=opts) as p:
    with beam_impl.Context(temp_dir=os.path.join(OUTPUT_DIR, 'tmp')):
      
      # Write the raw data metadata to disk
      # Without the overloaded operators: p.apply(tft_beam_io.WriteMetadata(os.path.join(OUTPUT_DIR, 'metadata/rawdata_metadata'), raw_data_metadata)
      _ = (raw_data_metadata
        | 'WriteInputMetadata' >> tft_beam_io.WriteMetadata(
            os.path.join(OUTPUT_DIR, 'metadata/rawdata_metadata'),
            pipeline=p))
           
      # Analyze and transform training data
      this_query = create_query(1, EVERY_N)
      
      # Read in training data from BigQuery table
      raw_data = (p
        # Get raw training data from BigQuery
        | 'train_read' >> beam.io.Read(beam.io.BigQuerySource(query=this_query, use_standard_sql=True))
        # Use our is_valid function to only retain valid examples from training data
        | 'train_filter' >> beam.Filter(is_valid))

      # Package raw training data and its metadata into a 'dataset'
      raw_dataset = (raw_data, raw_data_metadata)
      
      # Using the preprocessing function `preprocess_tft`, preprocess the training data
      # and produce a transformed training dataset and a function to transform other data later
      transformed_dataset, transform_fn = (
          raw_dataset | beam_impl.AnalyzeAndTransformDataset(preprocess_tft))
      
      # Break out the transformed training data and its metadata
      transformed_data, transformed_metadata = transformed_dataset
      
      # Write the transformed training data to files
      _ = transformed_data | 'WriteTrainData' >> tfrecordio.WriteToTFRecord(
          os.path.join(OUTPUT_DIR, 'train'),
          file_name_suffix='.gz',
          coder=example_proto_coder.ExampleProtoCoder(
              transformed_metadata.schema))
      
      # Read in test data from BigQuery table and filter as we did with training data
      raw_test_data = (p 
        | 'eval_read' >> beam.io.Read(beam.io.BigQuerySource(query=create_query(2, EVERY_N), use_standard_sql=True))
        | 'eval_filter' >> beam.Filter(is_valid))
      
      # Package test data and metadata into a dataset
      raw_test_dataset = (raw_test_data, raw_data_metadata)
      
      # Using the same transformation function that was calculated above, transform the test dataset
      transformed_test_dataset = (
          (raw_test_dataset, transform_fn) | beam_impl.TransformDataset())
      
      # Write the transformed test data to files
      transformed_test_data, _ = transformed_test_dataset
      _ = transformed_test_data | 'WriteTestData' >> tfrecordio.WriteToTFRecord(
          os.path.join(OUTPUT_DIR, 'eval'),
          file_name_suffix='.gz',
          coder=example_proto_coder.ExampleProtoCoder(
              transformed_metadata.schema))
      
      # Write the transformation function to a file, as well
      _ = (transform_fn
           | 'WriteTransformFn' >>
           transform_fn_io.WriteTransformFn(os.path.join(OUTPUT_DIR, 'metadata')))

# Preprocess the training/test data
preprocess(in_test_mode=True, EVERY_N=None)

Launching local job ... hang on
INFO:tensorflow:Assets added to graph.
INFO:tensorflow:No assets to write.
INFO:tensorflow:SavedModel written to: ./preproc_tft/tmp/tftransform_tmp/19bb5413bd4b4b0f8e36621e3922af0c/saved_model.pb
INFO:tensorflow:Assets added to graph.
INFO:tensorflow:No assets to write.
INFO:tensorflow:SavedModel written to: ./preproc_tft/tmp/tftransform_tmp/b65580123bf9446cb74310cb4a9f1128/saved_model.pb


/usr/local/envs/py2env/lib/python2.7/site-packages/apache_beam/runners/direct/direct_runner.py:342: DeprecationWarning: options is deprecated since First stable release.. References to <pipeline>.options will not be supported
  pipeline.replace_all(_get_transform_overrides(pipeline.options))


INFO:tensorflow:Saver not created because there are no variables in the graph to restore


INFO:tensorflow:Saver not created because there are no variables in the graph to restore


INFO:tensorflow:Saver not created because there are no variables in the graph to restore


INFO:tensorflow:Saver not created because there are no variables in the graph to restore


INFO:tensorflow:Assets added to graph.


INFO:tensorflow:Assets added to graph.


INFO:tensorflow:No assets to write.


INFO:tensorflow:No assets to write.


INFO:tensorflow:SavedModel written to: ./preproc_tft/tmp/tftransform_tmp/9ca1a4d8d57945c898e9ad464331262c/saved_model.pb


INFO:tensorflow:SavedModel written to: ./preproc_tft/tmp/tftransform_tmp/9ca1a4d8d57945c898e9ad464331262c/saved_model.pb


INFO:tensorflow:Saver not created because there are no variables in the graph to restore


INFO:tensorflow:Saver not created because there are no variables in the graph to restore


INFO:tensorflow:Saver not created because there are no variables in the graph to restore


INFO:tensorflow:Saver not created because there are no variables in the graph to restore


INFO:tensorflow:Saver not created because there are no variables in the graph to restore


INFO:tensorflow:Saver not created because there are no variables in the graph to restore
/usr/local/envs/py2env/lib/python2.7/site-packages/simplejson/encoder.py:286: DeprecationWarning: Interpreting naive datetime as local 2018-07-10 01:20:05.224311. Please add timezone info to timestamps.
  chunks = self.iterencode(o, _one_shot=True)


In [11]:
%bash
ls -ls preproc_tft
ls -ls preproc_tft/metadata
# gsutil ls -l gs://${BUCKET}/analysis/hallelujah-effect/preproc_tft/
# gsutil ls -l gs://${BUCKET}/analysis/hallelujah-effect/preproc_tft/metadata

total 24
 4 -rw-r--r-- 1 root root  3071 Jul 10 01:20 eval-00000-of-00001.gz
 4 drwxr-xr-x 5 root root  4096 Jul 10 01:20 metadata
 4 drwxr-xr-x 3 root root  4096 Jul 10 01:20 tmp
12 -rw-r--r-- 1 root root 11567 Jul 10 01:20 train-00000-of-00001.gz
total 12
4 drwxr-xr-x 3 root root 4096 Jul 10 01:20 rawdata_metadata
4 drwxr-xr-x 3 root root 4096 Jul 10 01:20 transformed_metadata
4 drwxr-xr-x 3 root root 4096 Jul 10 01:20 transform_fn


<h2> Train off preprocessed data </h2>

### Local Manual Training

In [25]:
MODEL_NAME = 'basic_features_hyperparameter_tuning'
os.environ['MODEL_NAME'] = MODEL_NAME

In [29]:
%%bash

OUTPUT_DIR=${PWD}/../models/${MODEL_NAME}
rm -rf ${OUTPUT_DIR}
export PYTHONPATH=${PYTHONPATH}:$PWD/trainer
python -m trainer.task \
   --train_data_paths="${PWD}/preproc_tft/train*" \
   --eval_data_paths="${PWD}/preproc_tft/eval*" \
   --train_steps=250 \
   --train_batch_size=${TRAIN_N} \
   --eval_steps=1 \
   --output_dir=${OUTPUT_DIR} \
   --job-dir=/tmp \
   --metadata_path="${PWD}/preproc_tft/metadata"

/usr/local/envs/py2env/lib/python2.7/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
INFO:tensorflow:Using config: {'_save_checkpoints_secs': 60, '_session_config': None, '_keep_checkpoint_max': 10, '_task_type': 'worker', '_train_distribute': None, '_is_chief': True, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7fd4146a9750>, '_evaluation_master': '', '_save_checkpoints_steps': None, '_keep_checkpoint_every_n_hours': 10000, '_service': None, '_num_ps_replicas': 0, '_tf_random_seed': None, '_master': '', '_num_worker_replicas': 1, '_task_id': 0, '_log_step_count_steps': 100, '_model_dir': '/content/datalab/notebooks/eim-analysis/Hallelujah_Effect/basic_features_hyperparameter_tuning/../models/basic_features_hyperparameter_tu

Results for 2500 batches of 303 cases (2500 training epochs using a `DNNClassifier`):

- Accuracy = 0.6721311
- Accuracy baseline = 0.6721312
- AUC = 0.5
- AUC-PR = 0.6639344
- Average loss = 0.64499277
- F1 score = 0.0
- False negatives = 20.0
- False positives = 0.0
- Label/mean = 0.32786885
- Loss = 39.34456
- Precision = 0.0
- Prediction/mean = 0.25737455
- Recall = 0.0
- True negatives = 41.0
- True positives = 0.0

### Local ML Engine Training

In [19]:
%%bash

OUTPUT_DIR=${PWD}/../models/${MODEL_NAME}
rm -rf ${OUTPUT_DIR}
gcloud ml-engine local train \
   --module-name=trainer.task \
   --package-path=${PWD}/trainer/trainer \
   --job-dir=$OUTPUT_DIR \
   -- \
   --train_data_paths="${PWD}/preproc_tft/train*" \
   --eval_data_paths="${PWD}/preproc_tft/eval*" \
   --train_steps=2500 \
   --train_batch_size=${TRAIN_N} \
   --eval_steps=1 \
   --output_dir=${OUTPUT_DIR} \
   --metadata_path="${PWD}/preproc_tft/metadata/"

/usr/local/envs/py2env/lib/python2.7/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
INFO:tensorflow:TF_CONFIG environment variable: {u'environment': u'cloud', u'cluster': {}, u'job': {u'args': [u'--train_data_paths=/content/datalab/notebooks/eim-analysis/Hallelujah_Effect/basic_features_hyperparameter_tuning/preproc_tft/train*', u'--eval_data_paths=/content/datalab/notebooks/eim-analysis/Hallelujah_Effect/basic_features_hyperparameter_tuning/preproc_tft/eval*', u'--train_steps=2500', u'--train_batch_size=303', u'--eval_steps=1', u'--output_dir=/content/datalab/notebooks/eim-analysis/Hallelujah_Effect/basic_features_hyperparameter_tuning/../models/basic_features_basic_dnn_classifier', u'--metadata_path=/content/datalab/notebooks/eim-analysis/Hallelujah_Effect

### Create hyper-parameter configuration

The file specifies the search region in parameter space.  Cloud MLE carries out a smart search algorithm within these constraints (i.e. it does not try out every single value).

In [46]:
%writefile hyperparam.yaml
trainingInput:
  scaleTier: STANDARD_1
  hyperparameters:
    goal: MAXIMIZE
    maxTrials: 1000
    maxParallelTrials: 5
    hyperparameterMetricTag: auc_precision_recall
    params:
    - parameterName: train_steps
      type: DISCRETE
      discreteValues: [500, 5000, 10000, 20000, 30000, 50000]
    - parameterName: hidden_units
      type: CATEGORICAL
      categoricalValues: ["128 32 4", "256 128 16", "64 64 64 8", "16 8 4"]
    - parameterName: dropout
      type: DISCRETE
      discreteValues: [0., 0.25, 0.5, 0.75]
    - parameterName: learning_rate
      type: DISCRETE
      discreteValues: [0.001, 0.01, 0.1, 0.25, 0.5]
    - parameterName: activation_function
      type: CATEGORICAL
      categoricalValues: ["relu", "leaky_relu", "elu"]

Overwriting hyperparam.yaml


### Cloud ML Engine Training

In [47]:
%%bash

OUTPUT_DIR=gs://${BUCKET}/analysis/hallelujah-effect/models/${MODEL_NAME}
JOBNAME=hallelujah_effect$(date -u +%y%m%d_%H%M%S)
echo ${OUTPUT_DIR} $REGION $JOBNAME
gsutil -m rm -rf ${OUTPUT_DIR}
gcloud ml-engine jobs submit training $JOBNAME \
   --region=$REGION \
   --package-path=${PWD}/trainer/trainer \
   --module-name=trainer.task \
   --job-dir=${OUTPUT_DIR} \
   --scale-tier=STANDARD_1 \
   --runtime-version=1.8 \
   --config=hyperparam.yaml \
   -- \
   --train_data_paths="gs://${BUCKET}/analysis/hallelujah-effect/preproc_tft/train*" \
   --eval_data_paths="gs://${BUCKET}/analysis/hallelujah-effect/preproc_tft/eval*" \
   --output_dir=$OUTPUT_DIR \
   --train_steps=25 \
   --train_batch_size=${TRAIN_N} \
   --eval_steps=1 \
   --metadata_path=gs://${BUCKET}/analysis/hallelujah-effect/preproc_tft/metadata/

gs://eim-muse/analysis/hallelujah-effect/models/basic_features_hyperparameter_tuning us-central1 hallelujah_effect180711_200028
jobId: hallelujah_effect180711_200028
state: QUEUED


Removing gs://eim-muse/analysis/hallelujah-effect/models/basic_features_hyperparameter_tuning/#1531338874873021...
Removing gs://eim-muse/analysis/hallelujah-effect/models/basic_features_hyperparameter_tuning/checkpoint#1531338879275020...
Removing gs://eim-muse/analysis/hallelujah-effect/models/basic_features_hyperparameter_tuning/eval/#1531338842417074...
Removing gs://eim-muse/analysis/hallelujah-effect/models/basic_features_hyperparameter_tuning/eval/events.out.tfevents.1531338842.cmle-training-master-e2f23860a9-0-bzdkm#1531338885953067...
Removing gs://eim-muse/analysis/hallelujah-effect/models/basic_features_hyperparameter_tuning/events.out.tfevents.1531338826.cmle-training-master-e2f23860a9-0-bzdkm#1531338895043860...
Removing gs://eim-muse/analysis/hallelujah-effect/models/basic_features_hyperparameter_tuning/eval/events.out.tfevents.1531339188.cmle-training-master-946141f238-0-pp84x#1531339190406430...
Removing gs://eim-muse/analysis/hallelujah-effect/models/basic_features_hyp

Results for 2500 batches of 303 cases (2500 training epochs using a `DNNClassifier`):

- Accuracy = 0.55737704
- Accuracy baseline = 0.6721312
- AUC = 0.4573171
- AUC-PR = 0.30257925
- Average loss = 13.65598
- F1 score = 0.12903225
- False negatives = 18.0
- False positives = 9.0
- Label/mean = 0.32786885
- Loss = 833.0148
- Precision = 0.18181819
- Prediction/mean = 0.190581
- Recall = 0.1
- True negatives = 32.0
- True positives = 2.0

## View Results in TensorBoard

In [16]:
from google.datalab.ml import TensorBoard
TensorBoard().start('gs://eim-muse/analysis/hallelujah-effect/models/')

TensorBoard was started successfully with pid 4344. Click here to access it.

4344

In [10]:
TensorBoard.stop(327)

In [ ]:
%writefile /tmp/test.json
{"age":"29.0","activity":3.0}

In [ ]:
%bash
model_dir=$(ls $PWD/hallelujah-effect_trained/export/exporter/)
gcloud ml-engine local predict \
    --model-dir=./hallelujah-effect_trained/export/exporter/${model_dir} \
    --json-instances=/tmp/test.json

# To Do

- Double-check that Dublin ratings are inverted properly
- LASSO to identify important features
- Hyperparameter search
- More plots and statistics from the dataset with which I'm working here
- Bring in rows with missing values
- Feature engineering (physiological signals, MIR, feature crosses, variable-width binning)
- Include signals with good quality only in reaction range
- Customize estimator to add additional metrics

In [22]:
print('gs://{}/analysis/hallelujah-effect/models/{}'.format(BUCKET, MODEL_NAME))

gs://eim-muse/analysis/hallelujah-effect/models/basic_features


/usr/local/envs/py2env/lib/python2.7/site-packages/simplejson/encoder.py:286: DeprecationWarning: Interpreting naive datetime as local 2018-07-01 20:34:08.418586. Please add timezone info to timestamps.
  chunks = self.iterencode(o, _one_shot=True)


In [17]:
%%bash

tensorboard --inspect --logdir gs://eim-muse/analysis/hallelujah-effect/models/basic_features

Processing event files... (this can take a few minutes)

Found event files in:
gs://eim-muse/analysis/hallelujah-effect/models/basic_features
gs://eim-muse/analysis/hallelujah-effect/models/basic_features/eval/

These tags are in gs://eim-muse/analysis/hallelujah-effect/models/basic_features:
audio -
histograms
   dnn/dnn/hiddenlayer_0/activation
   dnn/dnn/hiddenlayer_1/activation
   dnn/dnn/hiddenlayer_2/activation
   dnn/dnn/logits/activation
images -
scalars
   average_loss
   dnn/dnn/hiddenlayer_0/fraction_of_zero_values
   dnn/dnn/hiddenlayer_1/fraction_of_zero_values
   dnn/dnn/hiddenlayer_2/fraction_of_zero_values
   dnn/dnn/logits/fraction_of_zero_values
   loss
   read_batch_features/file_name_queue/fraction_of_32_full
   read_batch_features/fraction_over_9970_of_30_full
   read_batch_features/queue/parsed_features/read_batch_features/fifo_queue/fraction_of_100_full
tensor -

Event statistics for gs://eim-muse/analysis/hallelujah-effect/models/basic_features:
audio -
graph
  

/usr/local/envs/py2env/lib/python2.7/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


In [32]:
ml = """
This is
a
multiline string.
"""

with open('./test.txt', 'w') as file:
  file.write(ml)

/usr/local/envs/py2env/lib/python2.7/site-packages/simplejson/encoder.py:286: DeprecationWarning: Interpreting naive datetime as local 2018-07-04 00:58:42.778246. Please add timezone info to timestamps.
  chunks = self.iterencode(o, _one_shot=True)
